control-plane/agents/core/src/pool/mod.rs

```diff
@@ -146,4 +146,167 @@ mod tests {
+        .unwrap();
+
+        async fn check_operation(replica: &Replica, protocol: Protocol) {
+            // operation in progress
+            assert!(replica_spec(&replica).await.unwrap().operation.is_some());
+            tokio::time::delay_for(std::time::Duration::from_millis(500)).await;
+            // operation is completed
+            assert!(replica_spec(&replica).await.unwrap().operation.is_none());
+            assert_eq!(replica_spec(&replica).await.unwrap().share, protocol);
+        }
+
+        // pause mayastor
+        cluster.composer().pause(mayastor.as_str()).await.unwrap();
+
+        ShareReplica::from(&replica)
+            .request_ext(bus_timeout_opts())
+            .await
+            .expect_err("mayastor down");
+
+        check_operation(&replica, Protocol::Off).await;
+
+        // unpause mayastor
+        cluster.composer().thaw(mayastor.as_str()).await.unwrap();
+
+        // now it should be shared successfully
+        let uri = ShareReplica::from(&replica).request().await.unwrap();
+        println!("Share uri: {}", uri);
+
+        cluster.composer().pause(mayastor.as_str()).await.unwrap();
+
+        UnshareReplica::from(&replica)
+            .request_ext(bus_timeout_opts())
+            .await
+            .expect_err("mayastor down");
+
+        check_operation(&replica, Protocol::Nvmf).await;
+
+        cluster.composer().thaw(mayastor.as_str()).await.unwrap();
+
+        UnshareReplica::from(&replica).request().await.unwrap();
+
+        assert_eq!(replica_spec(&replica).await.unwrap().share, Protocol::Off);
+    }
+
+    #[actix_rt::test]
+    async fn replica_transaction_store() {
+        let store_timeout = Duration::from_millis(250);
+        let reconcile_period = Duration::from_millis(250);
+        let cluster = ClusterBuilder::builder()
+            .with_rest(false)
+            .with_pools(1)
+            .with_agents(vec!["core"])
+            .with_node_timeouts(
+                Duration::from_millis(500),
+                Duration::from_millis(500),
+            )
+            .with_reconcile_period(reconcile_period)
+            .with_store_timeout(store_timeout)
+            .with_bus_timeouts(bus_timeout_opts())
+            .build()
+            .await
+            .unwrap();
+        let mayastor = cluster.node(0);
+
+        let replica = CreateReplica {
+            node: mayastor.clone(),
+            uuid: ReplicaId::new(),
+            pool: cluster.pool(0, 0),
+            size: 12582912,
+            thin: false,
+            share: Protocol::Off,
+            ..Default::default()
+        }
+        .request()
+        .await
+        .unwrap();
+
+        // pause mayastor
+        cluster.composer().pause(mayastor.as_str()).await.unwrap();
+
+        ShareReplica::from(&replica)
+            .request_ext(bus_timeout_opts())
+            .await
+            .expect_err("mayastor down");
+
+        // swap them so that the share will succeed but etcd store will fail
```

Comment:
Maybe we can clarify this comment by explicitly saying that we are going to pause etcd and unpause mayastor.

Reply:
Done